### PR TITLE
Remove a couple of goroutine leaks in case of processing/context failures

### PR DESCRIPTION
**Pull request by gm42**
Linearize some error handling logic to avoid goroutine leaks.

Issue: #206

## What problem is this PR solving?

There are goroutine leaks at:
* https://github.com/beatlabs/patron/blob/3e49af8/component/async/component.go#L178
* https://github.com/beatlabs/patron/blob/3e49af8/component/async/component.go#L181

They also cause the Sarama errors channel and `failCh` to leak as a side-effect.

### CODE REVIEW

**Review by mantzas — DISMISSED** [review]
Nice catch! LGTM

**Comment by mantzas** [comment]
@gm42 is it possible to open also a corresponding issue and reference it in this PR?
The tests are also broken...

**Comment by gm42** [comment]
@mantzas I will open an issue; as for the tests: what approach do you prefer?
1. preserve 100% flow/errors as they are now
2. return the `ctx.Err()` (which may or not be `context.Canceled`) as an error in case `ctx.Done()` is selected

For now I'll go for (1) but I thought maybe we should document this behavior and/or log that error and/or return it.

File: component/async/component.go [hunk 1]
```diff
@@ -168,38 +168,32 @@ func (c *Component) processing(ctx context.Context) error {
...
-			case msg := <-chMsg:
-				log.Debug("New message from consumer arrived")
-				c.processMessage(msg, failCh)
-			case errMsg := <-chErr:
-				failCh <- fmt.Errorf("an error occurred during message consumption: %w", errMsg)
-				return
+	for {
+		select {
+		case msg := <-chMsg:
+			log.Debug("New message from consumer arrived")
```
Comment (oss92) [inline component/async/component.go]: maybe make this line too start with a lowercase letter and perhaps rephrase it to `consumer received a new message`. Also since it it the first start of log correlation in the consumer, perhaps `log.FromContext(msg.Context()).Debug()`?

Comment (mantzas) [reply component/async/component.go]: i agree

Comment (gm42) [reply component/async/component.go]: Sure, this was pre-existing code unrelated to this PR but I can improve it.

Shall I also add the message with `.Sub()` for structured logging purposes?

Comment (mantzas) [reply component/async/component.go]: There is no need for this.

Comment (gm42) [reply component/async/component.go]: Changed.

Comment (mantzas) [reply component/async/component.go]: I do not see the change actually... ;)

Comment (gm42) [reply component/async/component.go]: It's here: https://github.com/beatlabs/patron/pull/205/files#diff-6c64b2e2b8b9d28dcc79aa7f5ec20954R174

Comment (gm42) [reply component/async/component.go]: I will add the context too.

Comment (mantzas) [reply component/async/component.go]: you forgot to `log.FromContext(msg.Context()).Debug(...)`

Comment (gm42) [reply component/async/component.go]: Done.

Comment (mantzas) [reply component/async/component.go]: Sorry, but you changed all logging... The only log that needed to be chnaged was the mentioned one...
There is no context need for the others.

Comment (gm42) [reply component/async/component.go]: @mantzas sure, I changed them back.

I fixed a typo in the README as well.

**Review by mantzas — APPROVED** [review]

File: component/async/component.go [hunk 2]
```diff
@@ -168,38 +168,32 @@ func (c *Component) processing(ctx context.Context) error {
...
-			case errMsg := <-chErr:
-				failCh <- fmt.Errorf("an error occurred during message consumption: %w", errMsg)
-				return
+	for {
+		select {
+		case msg := <-chMsg:
+			log.FromContext(msg.Context()).Debug("consumer received a new message")
+			err := c.processMessage(msg)
+			if err != nil {
+				return err
```
Comment (drakos74) [inline component/async/component.go:177]: thats an interesting case ... Do we want to stop processing at the error ? Or just ignore ? Also it would be interesting to see how the retry will work. (for now it was only used for connection failures). What scenario i am thinking of is the consumer error during deserialising of a poisonous message. 
Having said that ... should we maybe leave this open in a property ? And the default behavior could be the one you are implementing now (e.g. fail fast)

**Review by drakos74 — COMMENTED** [review]
nice! Just a small comment on the failure assumptions ...

File: component/async/component.go (hunk 2 above)
Comment (gm42) [reply component/async/component.go:177]: The behavior did not change in this PR, it was doing the same before.

I did not even change the behavior on context error (we are ignoring the error there, see a couple lines above).

I would suggest to not change behavior in this PR but address it in a different PR; for example you could have a channel for processing messages and fail the whole thing only in case of non-processing errors.

Comment (mantzas) [reply component/async/component.go:177]: I agree not to chnage anything in this PR...
Since Kafka is a stream and not a queue the problems that you mentioned should be handled by the consumer code. No DLQ here I am afraid.
If we ignore an error then the partition iterator will move forward and the message will be "lost"...

Comment (drakos74) [reply component/async/component.go:177]: yeap good point... missed the lines above :) .
I see now the behavior does not change .

**Review by drakos74 — APPROVED** [review]

**Comment by gm42** [comment]
Any tips on how to improve the code coverage? Other than that I do not have to make other changes here.
